### ISSUE TITLE
Emscripten - Clang 

### ISSUE THREAD

**Issue by afraidofdark** [alugowski/poolSTL#30]
Hi, 
I have tested the replacement in my project and it works just fine on windows visual studio 2022 and android studio 2022.3.1 (ndk 26.1.109)

It works as expected, however with emscripten, the code below fails with the provided log.
version of poolSTL 0.3.4

I am not sure if you have tested it with emscripten, this stack trace is from google chrome. I'll keep looking in to it, I have provided the report in case you can provide more info.

```
#define POOLSTL_STD_SUPPLEMENT 1
#include <poolSTL/poolstl.hpp>
#include <vector>

int  main()
{
    // Create a vector of 10 numbers
    std::vector<int> numbers = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

    // Sum the numbers in parallel using std::reduce with std::execution::par
    int sum = std::reduce(std::execution::par, numbers.begin(), numbers.end());

	return sum;
}
```

```
$std::__2::__throw_system_error(int, char const*) @ system_error.cpp:293
$std::__2::thread::thread<void (task_thread_pool::task_thread_pool::*)(), task_thread_pool::task_thread_pool*, void>(void (task_thread_pool::task_thread_pool::*&&)(), task_thread_pool::task_thread_pool*&&) @ thread.h:252
$void std::__2::allocator<std::__2::thread>::construct[abi:ue170004]<std::__2::thread, void (task_thread_pool::task_thread_pool::*)(), task_thread_pool::task_thread_pool*>(std::__2::thread*, void (task_thread_pool::task_thread_pool::*&&)(), task_thread_pool::task_thread_pool*&&) @ allocator.h:167
$void std::__2::allocator_traits<std::__2::allocator<std::__2::thread>>::construct[abi:ue170004]<std::__2::thread, void (task_thread_pool::task_thread_pool::*)(), task_thread_pool::task_thread_pool*, void>(std::__2::allocator<std::__2::thread>&, std::__2::thread*, void (task_thread_pool::task_thread_pool::*&&)(), task_thread_pool::task_thread_pool*&&) @ allocator_traits.h:296
$void std::__2::vector<std::__2::thread, std::__2::allocator<std::__2::thread>>::__emplace_back_slow_path<void (task_thread_pool::task_thread_pool::*)(), task_thread_pool::task_thread_pool*>(void (task_thread_pool::task_thread_pool::*&&)(), task_thread_pool::task_thread_pool*&&) @ vector:1660
$std::__2::thread& std::__2::vector<std::__2::thread, std::__2::allocator<std::__2::thread>>::emplace_back<void (task_thread_pool::task_thread_pool::*)(), task_thread_pool::task_thread_pool*>(void (task_thread_pool::task_thread_pool::*&&)(), task_thread_pool::task_thread_pool*&&) @ vector:1681
$task_thread_pool::task_thread_pool::start_threads(unsigned int) @ poolstl.hpp:407
$task_thread_pool::task_thread_pool::task_thread_pool(unsigned int) @ poolstl.hpp:157
$std::__2::__shared_ptr_emplace<task_thread_pool::task_thread_pool, std::__2::allocator<task_thread_pool::task_thread_pool>>::__shared_ptr_emplace[abi:ue170004]<>(std::__2::allocator<task_thread_pool::task_thread_pool>) @ shared_ptr.h:302
$std::__2::shared_ptr<task_thread_pool::task_thread_pool> std::__2::allocate_shared[abi:ue170004]<task_thread_pool::task_thread_pool, std::__2::allocator<task_thread_pool::task_thread_pool>, void>(std::__2::allocator<task_thread_pool::task_thread_pool> const&) @ shared_ptr.h:1022
$std::__2::shared_ptr<task_thread_pool::task_thread_pool> std::__2::make_shared[abi:ue170004]<task_thread_pool::task_thread_pool, void>() @ shared_ptr.h:1031
$poolstl::execution::internal::get_default_pool()::'lambda'()::operator()() const @ poolstl.hpp:686
$decltype(std::declval<poolstl::execution::internal::get_default_pool()::'lambda'()>()()) std::__2::__invoke[abi:ue170004]<poolstl::execution::internal::get_default_pool()::'lambda'()>(poolstl::execution::internal::get_default_pool()::'lambda'()&&) @ invoke.h:340
$void std::__2::__call_once_param<std::__2::tuple<poolstl::execution::internal::get_default_pool()::'lambda'()&&>>::__execute[abi:ue170004]<>(std::__2::__tuple_indices<...>) @ mutex:632
$std::__2::__call_once_param<std::__2::tuple<poolstl::execution::internal::get_default_pool()::'lambda'()&&>>::operator()[abi:ue170004]() @ mutex:624
$void std::__2::__call_once_proxy[abi:ue170004]<std::__2::tuple<poolstl::execution::internal::get_default_pool()::'lambda'()&&>>(void*) @ mutex:660
$std::__2::__call_once(unsigned long volatile&, void*, void (*)(void*)) @ mutex.cpp:238
$void std::__2::call_once[abi:ue170004]<poolstl::execution::internal::get_default_pool()::'lambda'()>(std::__2::once_flag&, poolstl::execution::internal::get_default_pool()::'lambda'()&&) @ mutex:677
$poolstl::execution::internal::get_default_pool() @ poolstl.hpp:686
$poolstl::execution::parallel_policy::pool() const @ poolstl.hpp:730
$std::__2::vector<std::__2::future<int>, std::__2::allocator<std::__2::future<int>>> poolstl::internal::parallel_chunk_for_1<poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, int (*)(std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>, int, std::__2::plus<int>), int, int&, std::__2::plus<int>&>(poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>, int (*)(std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>, int, std::__2::plus<int>), int*, int, int&, std::__2::plus<int>&) @ poolstl.hpp:1698
$std::__2::enable_if<std::is_base_of<poolstl::execution::poolstl_policy, std::__2::remove_cv<std::__2::remove_reference<poolstl::execution::parallel_policy const&>::type>::type>::value, int>::type std::reduce<poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, int, std::__2::plus<int>>(poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>, int, std::__2::plus<int>) @ poolstl.hpp:2578
$std::__2::enable_if<std::is_base_of<poolstl::execution::poolstl_policy, std::__2::remove_cv<std::__2::remove_reference<poolstl::execution::parallel_policy const&>::type>::type>::value, int>::type std::reduce<poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, int>(poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>, int) @ poolstl.hpp:2594
$std::__2::enable_if<std::is_base_of<poolstl::execution::poolstl_policy, std::__2::remove_cv<std::__2::remove_reference<poolstl::execution::parallel_policy const&>::type>::type>::value, std::__2::iterator_traits<std::__2::__wrap_iter<int*>>::value_type>::type std::reduce<poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>>(poolstl::execution::parallel_policy const&, std::__2::__wrap_iter<int*>, std::__2::__wrap_iter<int*>) @ poolstl.hpp:2605
$ToolKit::Game::Frame(float) @ Game.cpp:36
```

**Comment by afraidofdark** [comment]
It appears that I forgotto add -pthreads.
It works just fine after I have added
"-pthreads" and emscripten spesific "-sPTHREAD_POOL_SIZE=4"

**Comment by alugowski** [comment]
Yes, clang requires `-pthread` (though Apple Clang does not).

@afraidofdark do you know of a good way to test with emscripten in GitHub Actions?

**Comment by afraidofdark** [comment]
> Yes, clang requires `-pthread` (though Apple Clang does not).
> 
> @afraidofdark do you know of a good way to test with emscripten in GitHub Actions?

A quick internet search brought up ready make actions that loads the emscripten, so compiling should be pretty easy but I am not sure if you can run tests without having explicit access to build server. I have used self-hosted runner, which run actions on my local machine so I can be sure that emscripten can popup a browser to run tests.

Self Hosted Runners ( Run actions on your machine so you can be sure test environment contains everything )
https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners

I have seen this guy running tests on .net ?
https://github.com/paulirwin/WasmNet/actions/runs/6647491182/job/18062967664

If I spot anything I'll be back here, or a seperate ticket can be created for testing emscripten.

**Comment by alugowski** [comment]
So it requires some fiddling, but it should be doable.

Just curious, are you experiencing any issues using poolSTL on Emscripten? Is performance as expected?

**Comment by alugowski** [comment]
I've added CI tests of Emscripten in https://github.com/alugowski/poolSTL/pull/33

Basic setup appears to be pretty easy. Emscripten provides a toolchain file that makes CMake do all the right things. It even configures the tests properly, so `ctest` will run out of the box via node. Any plain executables can be run with `node <target_name>`. Node v21 handles everything out of the box, older versions might require some experimental switches for pthreads to work.

Like you say, `-pthread` switch is required for both compiling and linking. I've updated the poolSTL CMakeLists.txt to do this, and updated installation instructions to point it out. Note it's *not* `-lpthread` like is required by regular Clang.

At least on node, `"-sPTHREAD_POOL_SIZE=4"` appears to have no effect. Speedup is identical (and good) with and without. Though the Emscripten docs suggest `-sPTHREAD_POOL_SIZE=navigator.hardwareConcurrency`.

Benchmarks on my laptop show the expected speedups. On the CI hosts some methods act as expected, some (like sort) shows massive speedup, and some show a slowdown. CI hosts are pretty limited, though, and I see forum threads about tuning emscripten threading, so it might be worth benchmarking your particular application to make sure you're seeing the speedup you expect.
